test/matrix/matrix.cpp

```diff
@@ -0,0 +1,782 @@
+        auto dev_result = Csr::create(this->exec);
+
+        mtx.ref->convert_to(ref_result.get());
+        mtx.dev->convert_to(dev_result.get());
+
+        GKO_ASSERT_MTX_NEAR(ref_result, dev_result, 0.0);
+        GKO_ASSERT_MTX_EQ_SPARSITY(ref_result, dev_result);
+    });
+}
+
+
+TYPED_TEST(Matrix, ConvertFromCsrIsEquivalentToRef)
+{
+    using TestConfig = typename TestFixture::Config;
+    using Mtx = typename TestFixture::Mtx;
+    using Csr =
+        gko::matrix::Csr<typename Mtx::value_type, typename Mtx::index_type>;
+    this->forall_matrix_data_scenarios([&](auto data) {
+        auto ref_src = Csr::create(this->ref);
+        auto dev_src = Csr::create(this->exec);
+        ref_src->read(data);
+        dev_src->read(data);
+        auto ref_result = TestConfig::create(this->ref, data.size);
+        auto dev_result = TestConfig::create(this->exec, data.size);
+
+        ref_src->convert_to(ref_result.get());
+        dev_src->convert_to(dev_result.get());
+
+        GKO_ASSERT_MTX_NEAR(ref_result, dev_result, 0.0);
+        GKO_ASSERT_MTX_EQ_SPARSITY(ref_result, dev_result);
+    });
+}
+
+
+TYPED_TEST(Matrix, ConvertToDenseIsEquivalentToRef)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using Dense = gko::matrix::Dense<typename Mtx::value_type>;
+    this->forall_matrix_scenarios([&](auto mtx) {
+        auto ref_result = Dense::create(this->ref);
+        auto dev_result = Dense::create(this->exec);
+
+        mtx.ref->convert_to(ref_result.get());
+        mtx.dev->convert_to(dev_result.get());
+
+        GKO_ASSERT_MTX_NEAR(ref_result, dev_result, 0.0);
+    });
+}
+
+
+TYPED_TEST(Matrix, ConvertFromDenseIsEquivalentToRef)
+{
+    using TestConfig = typename TestFixture::Config;
+    using Mtx = typename TestFixture::Mtx;
+    using Dense = gko::matrix::Dense<typename Mtx::value_type>;
+    this->template forall_matrix_data_scenarios([&](auto data) {
+        auto ref_src = Dense::create(this->ref);
+        auto dev_src = Dense::create(this->exec);
+        ref_src->read(data);
+        dev_src->read(data);
+        auto ref_result = TestConfig::create(this->ref, data.size);
+        auto dev_result = TestConfig::create(this->exec, data.size);
+
+        ref_src->convert_to(ref_result.get());
+        dev_src->convert_to(dev_result.get());
+
+        GKO_ASSERT_MTX_NEAR(ref_result, dev_result, 0.0);
+        GKO_ASSERT_MTX_EQ_SPARSITY(ref_result, dev_result);
+    });
+}
+
+
+TYPED_TEST(Matrix, ReadWriteRoundtrip)
+{
+    using TestConfig = typename TestFixture::Config;
+    using value_type = typename TestFixture::value_type;
+    using index_type = typename TestFixture::index_type;
+    this->forall_matrix_data_scenarios([&](auto data) {
+        auto new_mtx = TestConfig::create(this->exec, data.size);
+        gko::matrix_data<value_type, index_type> out_data;
+
+        new_mtx->read(data);
+        new_mtx->write(out_data);
+
+        if (!TestConfig::preserves_zeros()) {
```

Comment:
the preserve_zeros depends on the matrix format now. Should we move preserve_zeros is an option of the matrix?

Reply:
In the future yes